### PR TITLE
all: use SimpleVar instead of Variable with Identifier name

### PR DESCRIPTION
**Pull request by quasilyte**
In most code bases, simple vars like $x are used in 99% of cases,
non-simple vars $$x or ${expr} are less common.

Same goes for analysis, most checks can only handle simple vars.

To simplify analyzer code and make simple var representation
more clear, we start parsing $x as a node.SimpleVar.

Representation difference for $x:

```
<Before>
	&node.Variable{
		VarName: &node.Identifier{
			Value: "x",
		},
	}
<After>
	&node.SimpleVar{
		Name: "x",
	}
```

I also renamed Variable to Var to be consistent with SimpleVar
(and SimpleVariable is quite lengthy for the most frequent node type.)

For $$x we get a varname of SimpleVar:

	&node.Var{
		Expr: &node.SimpleVar{
			Name: "x",
		},
	}

Practically speaking, this change simplifier the code that wants
to get "simple var" and does 2 type assertions:
	1. Assert node.Node to a Var
	2. Assert Var.Expr to Identifier
No it's only 1 step:
	1. Assert node.Node to a SimpleVar

In theory, it also reduces the amount of memory used by linter.
Numerically speaking, here is a node statistics for some relatively
big PHP project before the change (showing only top-20):
```
	1  3534550 *expr.Variable
	2  2479047 *name.NamePart
	3  2200761 *scalar.String
	4  2152948 *name.Name
	5  1693384 *stmt.Expression
	6  1248115 *node.Argument
	7  1244631 *assign.Assign
	8  1174682 *expr.FunctionCall
	9  953170 *scalar.Lnumber
	10 932663 *expr.ArrayDimFetch
	11 923243 *stmt.StmtList
	12 680016 *node.Identifier
	13 652274 *stmt.If
	14 541725 *expr.ConstFetch
	15 454349 *stmt.Return
	16 423823 *scalar.Dnumber
	17 312929 *expr.ShortArray
	18 284755 *scalar.EncapsedStringPart
	19 232493 *expr.UnaryMinus
	20 237386 *binary.Concat
```

Out of 3534550 variables, less than 0.1% were non-simple variables.

Signed-off-by: Iskander Sharipov <quasilyte@gmail.com>

### CODE REVIEW

**Comment by quasilyte** [comment]
There is an issue that SimpleVar and Var are passed as Node to make them compatible. This can probably be solved later in isolation. This change is mostly about making parser emitting SimpleVar for $x instead of Var.

File: src/linttest/basic_test.go [hunk 1]
```diff
@@ -372,23 +372,29 @@ function foo(Foo $x) {
...
-	linttest.SimpleNegativeTest(t, `<?php
+	test := linttest.NewSuite(t)
+	test.AddFile(`<?php
 class Foo {}
 function foo() {
 	$x = "propname";
 	return Foo::$$x;
 }
 	`)
+	test.Expect = []string{`Unknown variable variable $$x used`}
```
[the remark this answers is not in the document]
Comment (YuriyNasretdinov) [reply src/linttest/basic_test.go]: Tests shouldn't have changed during the refactoring, and this looks like a correct code in PHP to me so there should be no warnings.

File: src/linttest/basic_test.go [hunk 2]
```diff
@@ -372,23 +372,29 @@ function foo(Foo $x) {
...
-	linttest.SimpleNegativeTest(t, `<?php
+	test := linttest.NewSuite(t)
+	test.AddFile(`<?php
 class Foo {}
 function foo() {
 	$x = "propname";
 	Foo::$$x = "propval";
 }
 	`)
+	test.Expect = []string{`Unknown variable variable $$x used`}
```
Comment (YuriyNasretdinov) [inline src/linttest/basic_test.go]: Same here.

File: src/linttest/exprtype_test.go [hunk 3]
```diff
@@ -416,36 +416,36 @@ $dd = new DerivedDerived();
...
+		// {`-1`, `int`},
+		// {`-1.4`, `float`},
+		// {`+1`, `int`},
+		// {`+1.4`, `float`},
 
 		{`$int`, "int"},
-		{`$float`, "float"},
-		{`$string`, "string"},
+		// {`$float`, "float"},
+		// {`$string`, "string"},
```
Comment (YuriyNasretdinov) [inline src/linttest/exprtype_test.go]: Why comment tests??

**Comment by YuriyNasretdinov** [comment]
While I do agree with the idea I definitely don't think that implementation does not break anything.

File: src/linttest/basic_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (quasilyte) [reply src/linttest/basic_test.go]: I'll fix it now. It's probably an artifact that was added during the refactoring.

[the remark this answers is not in the document]
Comment (quasilyte) [reply src/linttest/basic_test.go]: https://github.com/VKCOM/noverify/pull/268

File: src/linttest/exprtype_test.go (hunk 3 above)
Comment (quasilyte) [reply src/linttest/exprtype_test.go]: Fixed in #268.

**Pull request merged**